sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -243,19 +243,29 @@ Scheduler::GraphBuilder::getOrInsertMemObjRecord(const QueueImplPtr &Queue,
 void Scheduler::GraphBuilder::updateLeaves(const std::set<Command *> &Cmds,
                                            MemObjRecord *Record,
                                            access::mode AccessMode,
+                                           const MapOfDependentCmds &DependentCmdsOfNewCmd,
+                                           const QueueImplPtr &Queue,
                                            std::vector<Command *> &ToCleanUp) {
 
   const bool ReadOnlyReq = AccessMode == access::mode::read;
-  if (ReadOnlyReq)
-    return;
-
   for (Command *Cmd : Cmds) {
-    bool WasLeaf = Cmd->MLeafCounter > 0;
-    Cmd->MLeafCounter -= Record->MReadLeaves.remove(Cmd);
-    Cmd->MLeafCounter -= Record->MWriteLeaves.remove(Cmd);
-    if (WasLeaf && Cmd->readyForCleanup()) {
-      ToCleanUp.push_back(Cmd);
+    if (! ReadOnlyReq) {
+      bool WasLeaf = Cmd->MLeafCounter > 0;
+      Cmd->MLeafCounter -= Record->MReadLeaves.remove(Cmd);
+      Cmd->MLeafCounter -= Record->MWriteLeaves.remove(Cmd);
+      if (WasLeaf && Cmd->readyForCleanup()) {
+        ToCleanUp.push_back(Cmd);
```

Comment:
Just to capture what we discussed here. Instead of adding a new condition where we send commands to cleanup, I think a better solution would be excluding such dependencies from leaves (where both the newest node and its dependency are leaves of the same memory object) and let the usual cleanup path handle it. That can be done by adding a redundant dependency to the graph, the same way we're keeping graph width under control by imposing a leaf limit on each memory object.

That approach would also cover more cases: it's not needed that every dependency of the old command is covered by the new one for us to be able to clean it up. We should only care about how many memory objects the old command is a leaf for, other dependencies shouldn't be relevant.